sycl/include/CL/sycl/handler.hpp

```diff
@@ -91,11 +92,14 @@ static Arg member_ptr_helper(RetType (Func::*)(Arg));
 // template <typename RetType, typename Func>
 // static void member_ptr_helper(RetType (Func::*)());
 
-template <typename F>
-decltype(member_ptr_helper(&F::operator())) argument_helper(F);
+template <typename F, typename SuggestedArgType>
```

Comment:
What is `SuggestedArgType` used for?

Reply:
For generic lambda we cannot take the address of the non-static member function. Thus, we cannot calculate the type of the `Callable` parameter. It becomes the caller responsibility to pass the `SuggestedArgType` or `DefaultArgType` in case of generic `Callable` (generic lambda or similar)

Reply:
I was just surprised that it was not used here. So I guess it is used in some generic context, passing always a type, even when unused?

Reply:
Yes, you are right. It can be unused if the _Arg_ type can be calculated. But we need to add this parameter to both primary path and fallback path (when `decltype(member_ptr_helper(&F::operator())) is ill-formed` we take another overload where `SuggestedArgType` is used.) On the caller side we just don't know if the callee side would need the `SuggestedArgType` or not. So, you are right, we pass it always.